router/internal/expr/expr.go

```diff
@@ -0,0 +1,169 @@
+	Path string `expr:"path"`
+	// Query is the parsed query parameters
+	Query map[string]string `expr:"query"`
+}
+
+// RequestHeaders is the context for the headers object in expressions. A user can access the headers directly by
+// key and get the array representation of the header values or use the Get method to get the first value.
+type RequestHeaders map[string][]string
+
+// Get returns the first value associated with the given key (Exported).
+// For convenience, we export a function to make the work with headers case-insensitive.
+func (r RequestHeaders) Get(key string) string {
+	return http.Header(r).Get(key)
+}
+
+// LoadRequest loads the request object into the context.
+func (r *RequestRootContext) LoadRequest(req *http.Request) {
+	if req == nil {
+		return
+	}
+
+	r.Request.Header = RequestHeaders(req.Header)
+
+	m, _ := url.ParseQuery(req.URL.RawQuery)
+	qv := make(map[string]string, len(m))
+
+	for k := range m {
+		qv[k] = m.Get(k)
+	}
+
+	r.Request.URL = RequestURL{
+		Method: req.Method,
+		Scheme: req.URL.Scheme,
+		Host:   req.URL.Host,
+		Path:   req.URL.Path,
+		Query:  qv,
+	}
+}
+
+type RequestAuth struct {
+	IsAuthenticated bool           `expr:"isAuthenticated"`
+	Type            string         `expr:"type"`
+	Claims          map[string]any `expr:"claims"`
+	Scopes          []string       `expr:"scopes"`
+}
+
+// LoadAuth loads the authentication context into the request object.
+// Must only be called when the authentication was successful.
+func (r *RequestRootContext) LoadAuth(ctx context.Context) {
+	authCtx := authentication.FromContext(ctx)
+	if authCtx == nil {
+		return
+	}
+
+	r.Request.Auth.Type = authCtx.Authenticator()
+	r.Request.Auth.IsAuthenticated = true
+	r.Request.Auth.Claims = authCtx.Claims()
+	r.Request.Auth.Scopes = authCtx.Scopes()
+}
+
+// CompileBoolExpression compiles an expression and returns the program. It is used for expressions that return bool.
+// The exprContext is used to provide the context for the expression evaluation.
+func CompileBoolExpression(s string) (*vm.Program, error) {
+	v, err := expr.Compile(s, expr.Env(RequestRootContext{}), expr.AsBool())
+	if err != nil {
+		return nil, handleExpressionError(err)
+	}
+
+	return v, nil
+}
+
+// CompileStringExpression compiles an expression and returns the program. It is used for expressions that return strings
+// The exprContext is used to provide the context for the expression evaluation.
+func CompileStringExpression(s string) (*vm.Program, error) {
+	v, err := expr.Compile(s, expr.Env(RequestRootContext{}), expr.AsKind(reflect.String))
+	if err != nil {
+		return nil, handleExpressionError(err)
+	}
+
+	return v, nil
+}
+
+// ResolveStringExpression evaluates the expression and returns the result as a string. The exprContext is used to
+// provide the context for the expression evaluation.
+func ResolveStringExpression(vm *vm.Program, ctx RequestRootContext) (string, error) {
+	r, err := expr.Run(vm, ctx)
```

Comment:
Thread safe?

Reply:
Not required by the implementation.

Reply:
I revisited the implementation and you're right it is required but luckily `vm.program` is already thread safe.